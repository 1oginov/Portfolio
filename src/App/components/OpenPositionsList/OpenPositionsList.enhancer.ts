import { withProps } from 'recompose';

import { Props } from './OpenPositionsList';

export default withProps<Props, Props>(({ positions }) => ({
  positions: positions.filter(position => position.closeDate === null),
}));
