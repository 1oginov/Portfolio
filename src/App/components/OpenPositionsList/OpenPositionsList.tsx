import {
  List, ListItem, ListItemIcon, ListItemSecondaryAction, ListItemText,
} from '@material-ui/core';
import * as React from 'react';

import * as C from 'Constants';
import PositionDate from 'Portfolio/components/PositionDate';
import Money from 'Shared/components/Money';
import Percent from 'Shared/components/Percent';
import Sorter from 'Shared/components/Sorter';
import StockLogo from 'Stocks/components/StockLogo';

import { StockPosition } from '../../lib';

// TODO: Tests.

export interface Props {
  classes: { [key: string]: string };
  handleSorterKeyChange: (key: string) => void;
  handleSorterOrderChange: (order: 'asc' | 'desc') => void;
  onPositionClick?: (positionId: string) => void;
  sorterKey: string;
  sorterOrder: 'asc' | 'desc';
  stockPositions: StockPosition[];
  totalDailyPL: number;
  totalDailyPLPercent: number;
  totalMarketPL: number;
  totalMarketPLPercent: number;
  totalMarketSum: number;
}

const OpenPositionsList: React.FunctionComponent<Props> = ({
  classes, handleSorterKeyChange, handleSorterOrderChange, onPositionClick, stockPositions, sorterKey, sorterOrder,
  totalDailyPL, totalDailyPLPercent, totalMarketPL, totalMarketPLPercent, totalMarketSum,
}: Props) => (
  <React.Fragment>

    <Money value={totalMarketSum} />
    <Money pl value={totalMarketPL} />
    <Percent pl value={totalMarketPLPercent} />
    <Money pl value={totalDailyPL} />
    <Percent pl value={totalDailyPLPercent} />

    <Sorter
      keys={C.OPEN_POSITIONS_LIST_SORTER_KEYS}
      onKeyChange={handleSorterKeyChange}
      onOrderChange={handleSorterOrderChange}
      sorterKey={sorterKey}
      sorterOrder={sorterOrder}
    />

    <List dense className={classes.list}>
      {stockPositions.map(({
        amount, companyName, dailyPL, dailyPLPercent, id, marketPL, marketPLAnnualPercent, marketPLPercent, openDate,
        symbol,
      }) => (
        <ListItem button key={id} onClick={() => onPositionClick && onPositionClick(id)}>

          <ListItemIcon><StockLogo symbol={symbol} /></ListItemIcon>

          <ListItemText
            primary={companyName === null ? symbol : companyName}
            secondary={(
              <React.Fragment>
                {amount}
                {' @ '}
                <PositionDate highlighted={sorterKey === 'openDate'} date={openDate} />
              </React.Fragment>
            )}
          />

          {(sorterKey === 'dailyPL' || sorterKey === 'dailyPLPercent') && (
            <ListItemSecondaryAction className={classes.listItemSecondaryAction}>
              {dailyPL !== null && <Money highlighted={sorterKey === 'dailyPL'} pl value={dailyPL} />}
              <br />
              {dailyPLPercent !== null && (
                <Percent highlighted={sorterKey === 'dailyPLPercent'} pl value={dailyPLPercent} />
              )}
            </ListItemSecondaryAction>
          )}

          {sorterKey === 'marketPLAnnualPercent' && (
            <ListItemSecondaryAction className={classes.listItemSecondaryAction}>
              {marketPLAnnualPercent !== null && <Percent pl value={marketPLAnnualPercent} />}
            </ListItemSecondaryAction>
          )}

          {sorterKey !== 'dailyPL' && sorterKey !== 'dailyPLPercent' && sorterKey !== 'marketPLAnnualPercent' && (
            <ListItemSecondaryAction className={classes.listItemSecondaryAction}>
              {marketPL !== null && <Money highlighted={sorterKey === 'marketPL'} pl value={marketPL} />}
              <br />
              {marketPLPercent !== null && (
                <Percent highlighted={sorterKey === 'marketPLPercent'} pl value={marketPLPercent} />
              )}
            </ListItemSecondaryAction>
          )}

        </ListItem>
      ))}
    </List>

  </React.Fragment>
);

export default OpenPositionsList;
