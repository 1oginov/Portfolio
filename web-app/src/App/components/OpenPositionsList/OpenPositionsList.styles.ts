import { withStyles } from '@material-ui/core';

export default withStyles((theme) => ({
  list: {
    width: '100%',
  },
  listItemSecondaryAction: {
    textAlign: 'right',
  },
  sorter: {
    marginBottom: theme.spacing(2),
  },
}));
