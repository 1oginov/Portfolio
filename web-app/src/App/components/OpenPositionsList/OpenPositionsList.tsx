import {
  List, ListItem, ListItemIcon, ListItemSecondaryAction, ListItemText,
} from '@material-ui/core';
import * as React from 'react';

import * as C from 'Constants';
import { Money } from 'Layers/Presentation/Components/Money';
import { Percent } from 'Layers/Presentation/Components/Percent';
import { PositionDate } from 'Layers/Presentation/Components/PositionDate';
import { Sorter } from 'Layers/Presentation/Components/Sorter';
import { StockLogo } from 'Layers/Presentation/Components/StockLogo';

import { StockPosition } from '../../lib';

export interface Props {
  classes: { [key: string]: string };
  handleSorterKeyChange: (key: string) => void;
  handleSorterOrderChange: (order: 'asc' | 'desc') => void;
  onPositionClick?: (positionId: string) => void;
  sorterKey: string;
  sorterOrder: 'asc' | 'desc';
  stockPositions: StockPosition[];
}

const OpenPositionsList: React.FunctionComponent<Props> = ({
  classes, handleSorterKeyChange, handleSorterOrderChange, onPositionClick, stockPositions, sorterKey, sorterOrder,
}: Props) => (
  <>

    <Sorter
      className={classes.sorter}
      keys={C.OPEN_POSITIONS_LIST_SORTER_KEYS}
      onKeyChange={handleSorterKeyChange}
      onOrderChange={handleSorterOrderChange}
      sorterKey={sorterKey}
      sorterOrder={sorterOrder}
    />

    <List className={classes.list} dense disablePadding>
      {stockPositions.map(({
        amount, companyName, dailyPL, dailyPLPercent, id, marketPL, marketPLAnnualPercent, marketPLPercent, openDate,
        symbol,
      }) => (
        <ListItem button key={id} onClick={() => onPositionClick && onPositionClick(id)}>

          <ListItemIcon><StockLogo symbol={symbol} /></ListItemIcon>

          <ListItemText
            primary={companyName === null ? symbol : companyName}
            secondary={(
              <>
                {amount}
                {' '}
                @
                {' '}
                <PositionDate highlighted={sorterKey === 'openDate'} date={openDate} />
              </>
            )}
          />

          {(sorterKey === 'dailyPL' || sorterKey === 'dailyPLPercent') && (
            <ListItemSecondaryAction className={classes.listItemSecondaryAction}>
              {dailyPL !== null && <Money highlighted={sorterKey === 'dailyPL'} pl value={dailyPL} />}
              <br />
              {dailyPLPercent !== null && (
                <Percent highlighted={sorterKey === 'dailyPLPercent'} pl value={dailyPLPercent} />
              )}
            </ListItemSecondaryAction>
          )}

          {sorterKey === 'marketPLAnnualPercent' && (
            <ListItemSecondaryAction className={classes.listItemSecondaryAction}>
              {marketPLAnnualPercent !== null && <Percent pl value={marketPLAnnualPercent} />}
            </ListItemSecondaryAction>
          )}

          {sorterKey !== 'dailyPL' && sorterKey !== 'dailyPLPercent' && sorterKey !== 'marketPLAnnualPercent' && (
            <ListItemSecondaryAction className={classes.listItemSecondaryAction}>
              {marketPL !== null && <Money highlighted={sorterKey === 'marketPL'} pl value={marketPL} />}
              <br />
              {marketPLPercent !== null && (
                <Percent highlighted={sorterKey === 'marketPLPercent'} pl value={marketPLPercent} />
              )}
            </ListItemSecondaryAction>
          )}

        </ListItem>
      ))}
    </List>

  </>
);

export default OpenPositionsList;
