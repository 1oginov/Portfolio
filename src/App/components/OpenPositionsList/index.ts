import OpenPositionsList from './OpenPositionsList';

export default OpenPositionsList;
