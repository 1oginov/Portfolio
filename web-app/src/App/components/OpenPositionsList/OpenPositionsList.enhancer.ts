import { compose, withProps } from 'recompose';

import * as C from 'Constants';
import { Position } from 'Layers/Business/Services/PortfolioService/Interfaces/Position';
import { sortCollection } from 'Shared/lib';

import withSorter, { Props as WithSorterProps } from '../../enhancers/withSorter';
import { Props as WithStockPositionsProps } from '../../enhancers/withStockPositions';
import { StockPosition } from '../../lib';
import { Props } from './OpenPositionsList';

interface EnhancedProps {
  onPositionClick?: (positionId: string) => void;
  stockPositions: Position[];
}

export default compose<Props, EnhancedProps>(
  withProps<EnhancedProps, EnhancedProps>(({ stockPositions }) => ({
    stockPositions: stockPositions.filter((position) => position.closeDate === null),
  })),
  withSorter(
    C.OPEN_POSITIONS_LIST_SORTER_NAME,
    C.OPEN_POSITIONS_LIST_SORTER_INITIAL_KEY,
    C.OPEN_POSITIONS_LIST_SORTER_INITIAL_ORDER,
  ),
  withProps<Partial<Props>, EnhancedProps & WithStockPositionsProps & WithSorterProps>(({
    stockPositions, sorterKey, sorterOrder,
  }) => ({
    stockPositions: sortCollection(stockPositions, sorterKey as keyof StockPosition, sorterOrder),
  })),
);
