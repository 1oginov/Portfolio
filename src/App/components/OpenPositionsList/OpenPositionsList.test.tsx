import { shallow } from 'enzyme';
import classes from 'identity-obj-proxy';
import * as React from 'react';

import OpenPositionsList from './OpenPositionsList';

it('matches snapshot', () => {
  const wrapper = shallow(
    <OpenPositionsList
      classes={classes}
      handleSorterKeyChange={() => undefined}
      handleSorterOrderChange={() => undefined}
      sorterKey="openPrice"
      sorterOrder="asc"
      stockPositions={[]}
      totalDailyPL={10}
      totalDailyPLPercent={0.1}
      totalMarketPL={100}
      totalMarketPLPercent={1}
      totalMarketSum={1000}
    />,
  );

  expect(wrapper).toMatchSnapshot();
});
