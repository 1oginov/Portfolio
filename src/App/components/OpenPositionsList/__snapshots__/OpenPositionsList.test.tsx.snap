// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Fragment>
  <withState(withProps(Sorter))
    keys={
      Array [
        Object {
          "key": "symbol",
          "label": "Symbol",
        },
        Object {
          "key": "openDate",
          "label": "Open Date",
        },
        Object {
          "key": "dailyPL",
          "label": "Daily PL",
        },
        Object {
          "key": "dailyPLPercent",
          "label": "Daily PL%",
        },
        Object {
          "key": "marketPL",
          "label": "PL",
        },
        Object {
          "key": "marketPLPercent",
          "label": "PL%",
        },
        Object {
          "key": "marketPLAnnualPercent",
          "label": "Annual PL%",
        },
      ]
    }
    onKeyChange={[Function]}
    onOrderChange={[Function]}
    sorterKey="openPrice"
    sorterOrder="asc"
  />
  <WithStyles(ForwardRef(List))
    className="list"
    dense={true}
  />
</Fragment>
`;
